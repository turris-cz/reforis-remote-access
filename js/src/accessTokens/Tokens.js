/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import { useAPIGet, API_STATE } from "foris";

import API_URLs from "API";
import TokensTable from "./TokensTable";
import { useRevokeToken, useGenerateToken } from "./hooks";

Tokens.propTypes = {
    ws: PropTypes.object.isRequired,
    setGenerating: PropTypes.func.isRequired,
};

export default function Tokens({ ws, setGenerating }) {
    const [getTokensResponse, getTokens] = useAPIGet(API_URLs.tokens);
    useEffect(() => {
        getTokens();
    }, [getTokens]);

    const [tokens, setTokens] = useState([]);
    // Update list of tokens when GET request is successful
    useEffect(() => {
        if (getTokensResponse.state === API_STATE.SUCCESS) {
            setTokens(getTokensResponse.data);
        }
    }, [getTokensResponse]);

    // Refresh list of tokens when one is being generated
    useGenerateToken(ws, getTokens, setGenerating);

    // Refresh list of clients after certificate is revoked
    useRevokeToken(ws, setTokens);

    return (
        <>
            <h3>{_("Tokens")}</h3>
            <TokensTable
                apiState={getTokensResponse.state}
                tokens={tokens}
            />
        </>
    );
}
