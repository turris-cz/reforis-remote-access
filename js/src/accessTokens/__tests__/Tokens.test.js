/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    wait,
    act,
    getByText,
    getAllByText,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import { WebSockets } from "foris";
import mockAxios from "jest-mock-axios";

import { tokensFixture } from "./__fixtures__/tokens";
import Tokens from "../Tokens";

describe("<Tokens />", () => {
    let webSockets;
    const setGenerating = jest.fn();
    let container;

    beforeEach(() => {
        webSockets = new WebSockets();
        ({ container } = render(
            <Tokens ws={webSockets} setGenerating={setGenerating} />
        ));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render error", async () => {
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/remote-access/api/tokens",
            expect.anything()
        );
        mockJSONError();
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should render tokens", async () => {
        mockAxios.mockResponse({ data: tokensFixture });
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should refresh table after token is revoked", async () => {
        const getRevokedNumber = () =>
            getAllByText(container, "Access revoked").length;

        mockAxios.mockResponse({ data: tokensFixture });
        await wait(() => expect(getByText(container, "foo")).toBeTruthy());

        const revoked = getRevokedNumber();
        // Client is revoked
        act(() =>
            webSockets.dispatch({
                module: "remote",
                action: "revoke",
                data: { id: tokensFixture[0].id },
            })
        );
        expect(getRevokedNumber()).toBe(revoked + 1);
    });

    it("should refresh table after token is created", async () => {
        expect(mockAxios.get).toHaveBeenCalledTimes(1);
        mockAxios.mockResponse({ data: tokensFixture });
        await wait(() => expect(getByText(container, "foo")).toBeTruthy());

        const generatingMessage = {
            module: "remote",
            action: "generate_token",
            data: { status: "token_generating" },
        };
        act(() => webSockets.dispatch(generatingMessage));
        expect(mockAxios.get).toHaveBeenCalledTimes(2);

        const succeededMessage = {
            module: "remote",
            action: "generate_token",
            data: { status: "succeeded" },
        };
        act(() => webSockets.dispatch(succeededMessage));
        expect(mockAxios.get).toHaveBeenCalledTimes(3);

        expect(setGenerating).toBeCalledWith(false);
    });
});
