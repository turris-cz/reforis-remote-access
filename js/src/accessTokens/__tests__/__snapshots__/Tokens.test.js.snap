// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tokens /> should render error 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Tokens
    </h2>
    <p
      class="text-center text-danger"
    >
      An error occurred while fetching data.
    </p>
  </div>
</div>
`;

exports[`<Tokens /> should render spinner 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Tokens
    </h2>
    <div
      class="spinner-wrapper my-3 text-center"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tokens /> should render tokens 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Tokens
    </h2>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover mb-0"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              scope="col"
            >
              Name
            </th>
            <th
              aria-label="Actions"
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle"
            >
              foo
            </td>
            <td
              class="text-end"
            >
              <div
                class="btn-group"
                role="group"
              >
                <a
                  class="btn btn-primary btn-sm"
                  download=""
                  href="/reforis/remote-access/api/tokens/1"
                >
                  Download
                </a>
                <button
                  class="btn btn-danger btn-sm d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    Revoke
                  </span>
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="align-middle"
            >
              bar
            </td>
            <td
              class="text-end"
            >
              <span
                class="text-muted"
              >
                Access revoked
              </span>
            </td>
          </tr>
          <tr>
            <td
              class="align-middle"
            >
              baz
            </td>
            <td
              class="text-end"
            >
              <div
                class="spinner-border spinner-border-sm"
                role="status"
              >
                <span
                  class="sr-only"
                />
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
